docs/api_requests/block-vhost-user.md

```diff
@@ -0,0 +1,189 @@
+encryption or using sophisticated readahead logic. In such cases,
+the performance improvement will be coming from the fact that the custom logic
+is implemented in the same process that handles Virtio queues, which reduces
+the number of required context switches.
+
+## Backends
+
+There are a number of open source implementations of a vhost-user backend
+available for reference:
+
+1. [Qemu backend](https://github.com/qemu/qemu/tree/master/contrib/vhost-user-blk)
+1. [Cloud Hypervisor backend](https://github.com/cloud-hypervisor/cloud-hypervisor/tree/main/vhost_user_block)
+1. [crosvm backend](https://github.com/google/crosvm/blob/main/devices/src/virtio/vhost/user/device/block.rs)
+1. [SPDK backend](https://github.com/spdk/spdk/blob/master/lib/vhost/vhost_blk.c)
+
+## Security considerations
+
+### Guest memory sharing
+
+By design, a vhost-user frontend must share file descriptors
+of all guest memory regions to the backend. In order to achive that,
+guest memory is created as a [memfd](https://man7.org/linux/man-pages/man2/memfd_create.2.html)
+and mapped as `MAP_SHARED`.
+
+#### File descriptor in procfs
+
+An open `memfd` is reflected in `procfs` as any other open file descriptor:
+
+```shell
+$ ls -l /proc/{pid}/fd | grep memfd
+lrwx------ 1 1234 1234 64 Nov  2 13:39 32 -> /memfd:guest_mem (deleted)
+```
+
+Any process on the host that has access to this file in `procfs` will be able
+to map the file descriptor and observe runtime behaviour of the guest.
+
+At the moment, Firecracker does not close the `memfd`, because it must remain open
+until all the configured vhost-user devices have been activated and their info
+shared with the backends. This kind of tracking is not implemented in Firecracker,
+but may be implemented in the future. Meanwhile, users need to make sure that
+the access to the Firecracker's `procfs` tree is restricted to trusted processes
+on the host.
+
+On the backend side, it is advised that the backend closes
+the guest memory region file descriptors after mapping them into the address
+space.
+
+#### Resource limit in jailer
+
+The Firecracker [jailer](../jailer.md) allows to configure resource limits
+for the Firecracker process. Specifically, it allows to set the maximum file
+size. Since `memfd` that is used to back the guest memory is considered a file,
+the file size resource limit cannot be less than the biggest guest memory
+region. This does not require any special action from a user, but needs to be
+taken into consideration.
+
+### Remote code execution in the backend
+
+It is recommended to run Firecracker using the [jailer](../jailer.md). Since
+the vhost-user backend interacts with the guest via a Virtio queue, there is
+a potential for the guest to exercise issues in the backend code base
+to trigger undesired behaviours. Users should consider running their backend
+in a jailer or applying other adequate security measures to restrict it.
```

Comment:
maybe we should clarify that our jailer is not designed to jail arbitrary binaries, so users should use their own solutions (?)

Reply:
done